src/file-events/cpp/win_fsnotifier.cpp

```diff
@@ -231,9 +241,46 @@ Server::~Server() {
 void Server::runLoop(JNIEnv* env, function<void(exception_ptr)> notifyStarted) {
     notifyStarted(nullptr);
 
-    while (!terminated || watchPoints.size() > 0) {
+    while (!terminated) {
         SleepEx(INFINITE, true);
     }
+
+    // We have received termination, cancel all watchers
+    log_fine(env, "Finished with run loop, now cancelling remaining watch points", NULL);
+    int pendingWatchPoints = 0;
+    for (auto& it : watchPoints) {
+        auto& watchPoint = it.second;
+        switch (watchPoint.status) {
+            case LISTENING:
+                watchPoint.cancel();
+                pendingWatchPoints++;
+                break;
+            case CANCELLED:
+                pendingWatchPoints++;
+                break;
+            default:
+                break;
+        }
+    }
+
+    // If there are any pending watchers, wait for them to finish
+    if (pendingWatchPoints > 0) {
+        log_fine(env, "Waiting for %d pending watch points to finish", pendingWatchPoints);
+        SleepEx(SERVER_CLOSE_TIMEOUT_IN_MS, true);
```

Comment:
Same here, I don't think one round is necessarily enough to receive all the `ERROR_OPERATION_ABORTED` events.

Reply:
This is actually fine, as all the `CancelIoEx()` happen on the same thread, right before calling `SleepEx()`. So all the events are handled in one go, provided there is enough time in the timeout (for which a whole second should do).

Reply:
Seems like `0` time is also fine for this.